# Introduction to Data Science (CS4661). 
## Homework3

#### By: Harshil Patel(306646748)
---------------------------------------------------------------------------------------------------
---------------------------------------------------------------------------------------------------

### Question1: Predicting Heart Disease
In this question, we work with a dataset from the great textbook of "An Introduction to Statistical
Learning."

### a- 
Read the data file “Hearts_s.csv” (from github using the following command), and assign it to a Pandas DataFrame: df = pd.read_csv("https://github.com/mpourhoma/CS4661/raw/master/Heart_s.csv")

In [5]:
# The following line will import LogisticRegression and DecisionTreeClassifier Classes

from sklearn.linear_model import LogisticRegression

from sklearn.tree import DecisionTreeClassifier

from sklearn.neighbors import KNeighborsClassifier

In [6]:
# Importing the required packages and libraries
# we will need numpy and pandas later
import numpy as np
import pandas as pd

In [7]:
# reading a CSV file directly from Web, and store it in a pandas DataFrame:
# "read_csv" is a pandas function to read csv files from web or local device:

df = pd.read_csv("https://github.com/mpourhoma/CS4661/raw/master/Heart_s.csv")

In [8]:
# displaying the DataFrame:

df  # you can also use print(df)

Age Gender     ChestPain  RestBP  Chol  RestECG  MaxHR  Oldpeak  \
0     63      f       typical     145   233        2    150      2.3   
1     67      f  asymptomatic     160   286        2    108      1.5   
2     67      f  asymptomatic     120   229        2    129      2.6   
3     37      f    nonanginal     130   250        0    187      3.5   
4     41      m    nontypical     130   204        2    172      1.4   
5     56      f    nontypical     120   236        0    178      0.8   
6     62      m  asymptomatic     140   268        2    160      3.6   
7     57      m  asymptomatic     120   354        0    163      0.6   
8     63      f  asymptomatic     130   254        2    147      1.4   
9     53      f  asymptomatic     140   203        2    155      3.1   
10    57      f  asymptomatic     140   192        0    148      0.4   
11    56      m    nontypical     140   294        2    153      1.3   
12    56      f    nonanginal     130   256        2    142      0.6   
13    44      f    nontypical     120   263        0    173      0.0   
14    52      f    nonanginal     172   199        0    162      0.5   
15    57      f    nonanginal     150   168        0    174      1.6   
16    48      f    nontypical     110   229        0    168      1.0   
17    54      f  asymptomatic     140   239        0    160      1.2   
18    48      m    nonanginal     130   275        0    139      0.2   
19    49      f    nontypical     130   266        0    171      0.6   
20    64      f       typical     110   211        2    144      1.8   
21    58      m       typical     150   283        2    162      1.0   
22    58      f    nontypical     120   284        2    160      1.8   
23    58      f    nonanginal     132   224        2    173      3.2   
24    60      f  asymptomatic     130   206        2    132      2.4   
25    50      m    nonanginal     120   219        0    158      1.6   
26    58      m    nonanginal     120   340        0    172      0.0   
27    66      m       typical     150   226        0    114      2.6   
28    43      f  asymptomatic     150   247        0    171      1.5   
29    40      f  asymptomatic     110   167        2    114      2.0   
..   ...    ...           ...     ...   ...      ...    ...      ...   
271   71      m  asymptomatic     112   149        0    125      1.6   
272   59      f       typical     134   204        0    162      0.8   
273   64      f       typical     170   227        2    155      0.6   
274   66      m    nonanginal     146   278        2    152      0.0   
275   39      m    nonanginal     138   220        0    152      0.0   
276   57      f    nontypical     154   232        2    164      0.0   
277   58      m  asymptomatic     130   197        0    131      0.6   
278   57      f  asymptomatic     110   335        0    143      3.0   
279   47      f    nonanginal     130   253        0    179      0.0   
280   55      m  asymptomatic     128   205        1    130      2.0   
281   35      f    nontypical     122   192        0    174      0.0   
282   61      f  asymptomatic     148   203        0    161      0.0   
283   58      f  asymptomatic     114   318        1    140      4.4   
284   58      m  asymptomatic     170   225        2    146      2.8   
285   58      f    nontypical     125   220        0    144      0.4   
286   56      f    nontypical     130   221        2    163      0.0   
287   56      f    nontypical     120   240        0    169      0.0   
288   67      f    nonanginal     152   212        2    150      0.8   
289   55      m    nontypical     132   342        0    166      1.2   
290   44      f  asymptomatic     120   169        0    144      2.8   
291   63      f  asymptomatic     140   187        2    144      4.0   
292   63      m  asymptomatic     124   197        0    136      0.0   
293   41      f    nontypical     120   157        0    182      0.0   
294   59      f  asymptomatic     164   176  

### b- 
Check out the dataset. As you see, the dataset contains a number of features including both contextual and biological factors (e.g. age, gender, vital signs, …). The last column “AHD” is the label with “Yes” meaning that a human subject has Heart Disease, and “No” meaning that the subject does not have Heart Disease.

In [9]:
# select a Series of labels (the last column) from the DataFrame
y = df['AHD']
# checking the label vector by printing every  values
y[::1]

0       No
1      Yes
2      Yes
3       No
4       No
5       No
6      Yes
7       No
8      Yes
9      Yes
10      No
11      No
12     Yes
13      No
14      No
15      No
16     Yes
17      No
18      No
19      No
20      No
21      No
22     Yes
23     Yes
24     Yes
25      No
26      No
27      No
28      No
29     Yes
      ... 
271     No
272    Yes
273     No
274     No
275     No
276    Yes
277     No
278    Yes
279     No
280    Yes
281     No
282    Yes
283    Yes
284    Yes
285     No
286     No
287     No
288    Yes
289     No
290    Yes
291    Yes
292    Yes
293     No
294    Yes
295    Yes
296    Yes
297    Yes
298    Yes
299    Yes
300     No
Name: AHD, Length: 301, dtype: object

### c- 
As you see, there are at least 3 categorical features in the dataset (Gender, ChestPain,
Thal). Let’s ignore these categorical features for now, only keep the numerical features
and build your feature matrix and label vector.

In [10]:
# Creating the Feature Matrix for iris dataset:

# create a python list of feature names that would like to pick from the dataset:
feature_cols = ['Age','RestBP','Chol','RestECG','MaxHR','Oldpeak']

# use the above list to select the features from the original DataFrame
X = df[feature_cols]  

# checking the label vector by printing every  values
X[::1]

Age  RestBP  Chol  RestECG  MaxHR  Oldpeak
0     63     145   233        2    150      2.3
1     67     160   286        2    108      1.5
2     67     120   229        2    129      2.6
3     37     130   250        0    187      3.5
4     41     130   204        2    172      1.4
5     56     120   236        0    178      0.8
6     62     140   268        2    160      3.6
7     57     120   354        0    163      0.6
8     63     130   254        2    147      1.4
9     53     140   203        2    155      3.1
10    57     140   192        0    148      0.4
11    56     140   294        2    153      1.3
12    56     130   256        2    142      0.6
13    44     120   263        0    173      0.0
14    52     172   199        0    162      0.5
15    57     150   168        0    174      1.6
16    48     110   229        0    168      1.0
17    54     140   239        0    160      1.2
18    48     130   275        0    139      0.2
19    49     130   266        0    171      0.6
20    64     110   211        2    144      1.8
21    58     150   283        2    162      1.0
22    58     120   284        2    160      1.8
23    58     132   224        2    173      3.2
24    60     130   206        2    132      2.4
25    50     120   219        0    158      1.6
26    58     120   340        0    172      0.0
27    66     150   226        0    114      2.6
28    43     150   247        0    171      1.5
29    40     110   167        2    114      2.0
..   ...     ...   ...      ...    ...      ...
271   71     112   149        0    125      1.6
272   59     134   204        0    162      0.8
273   64     170   227        2    155      0.6
274   66     146   278        2    152      0.0
275   39     138   220        0    152      0.0
276   57     154   232        2    164      0.0
277   58     130   197        0    131      0.6
278   57     110   335        0    143      3.0
279   47     130   253        0    179      0.0
280   55     128   205        1    130      2.0
281   35     122   192        0    174      0.0
282   61     148   203        0    161      0.0
283   58     114   318        1    140      4.4
284   58     170   225        2    146      2.8
285   58     125   220        0    144      0.4
286   56     130   221        2    163      0.0
287   56     120   240        0    169      0.0
288   67     152   212        2    150      0.8
289   55     132   342        0    166      1.2
290   44     120   169        0    144      2.8
291   63     140   187        2    144      4.0
292   63     124   197        0    136      0.0
293   41     120   157        0    182      0.0
294   59     164   176        2     90      1.0
295   57     140   241        0    123      0.2
296   45     110   264        0    132      1.2
297   68     144   193        0    141      3.4
298   57     130   131        0    115      1.2
299   57     130   236        2    174      0.0
300   38     138   175        0    173      0.0

[301 rows x 6 columns]

### d- 
Split the dataset into testing and training sets with the following parameters:
test_size=0.25, random_state=6.

In [11]:
# Randomly splitting the original dataset into training set and testing set
# The function"train_test_split" from "sklearn.cross_validation" library performs random splitting.
# "test_size=0.3" means that pick 30% of data samples for testing set, and the rest (70%) for training set.

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, 
                                                    random_state=6)

In [12]:
# print the size of the traning set:
print(X_train.shape)
print(y_train.shape)

(225, 6)
(225,)


In [13]:
# print the size of the testing set:
print(X_test.shape)
print(y_test.shape)


(76, 6)
(76,)


### e- 
Use KNN (with k=3), Decision Tree (with random_state=5), and Logistic Regression
Classifiers to predict Heart Disease based on the training/testing datasets that you built
in part (d). Then check, compare, and report the accuracy of these 3 classifiers. Which one
is the best? Which one is the worst?

In [14]:
#importing KNeighborsClassifier class from sklearn.neighbors
from sklearn.neighbors import KNeighborsClassifier
#importing accuracy_score class from sklearn.metrics
from sklearn.metrics import accuracy_score

#declare the value of k
k = 3

#instantiating an object of KNeighborsClassifier with above k value
knn_modal = KNeighborsClassifier(n_neighbors=k)

#training the model using the training data and training labels
knn_modal.fit(X_train, y_train)

#predecting the testing data
y_predict = knn_modal.predict(X_test)

#calculating the accuracy of the modal
accuracy = accuracy_score(y_test, y_predict)

print(accuracy)

0.6447368421052632


In [15]:
#Logistic Regression

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=4)

my_logreg = LogisticRegression()
# Training ONLY on the training set:
my_logreg.fit(X_train, y_train)
y_predict_lr = my_logreg.predict(X_test)
score_lr = accuracy_score(y_test, y_predict_lr)
print(score_lr)

0.7236842105263158


C:\Users\patel\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


In [16]:
#Decision Tree (with random_state=5)

my_decisiontree = DecisionTreeClassifier(random_state=5)
# Training ONLY on the training set:
my_decisiontree.fit(X_train, y_train)
y_predict_dt = my_decisiontree.predict(X_test)
score_dt = accuracy_score(y_test, y_predict_dt)
print(score_dt)

0.6052631578947368


Which one is the best? Which one is the worst?
#Logistic Regression is the best(0.7236842105263158)
#Decision Tree is the worst(0.6052631578947368)

### f- 
Now, we want to use the categorical features as well! To this end, we have to perform a
feature engineering process called OneHotEncoding for the categorical features. To do
this, each categorical feature should be replaced with dummy columns in the feature
table (one column for each possible value of a categorical feature), and then encode it in
a binary manner such that only one of the dummy columns can take “1” at a time (and
zero for the rest). For example, “Gender” can take two values “m” and “f”. Thus, we need
to replace this feature (in the feature table) by 2 columns titled “m” and “f”. Wherever
we have a male subject, we can put “1” and ”0” in the columns “m” and “f”. Wherever
we have a female subject, we can put “0” and ”1” in the columns “m” and “f”. (Hint: you
will need 4 columns to encode “ChestPain” and 3 columns to encode “Thal”).

In [20]:
#making a list of the features that are needed
features = ['Age','Gender','ChestPain','RestBP','Chol','RestECG','MaxHR','Oldpeak','Thal']

#selecting above features from the data
X = df[feature_cols]

X= pd.get_dummies(X)
X.head()

Age  RestBP  Chol  RestECG  MaxHR  Oldpeak
0   63     145   233        2    150      2.3
1   67     160   286        2    108      1.5
2   67     120   229        2    129      2.6
3   37     130   250        0    187      3.5
4   41     130   204        2    172      1.4

### g- 
Repeat parts (d) and (e) with the new dataset that you built in part (f). How does the
prediction accuracy change for each method?

In [21]:
#spliting the data randomly into training and testing data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=4)

#importing KNeighborsClassifier class from sklearn.neighbors
from sklearn.neighbors import KNeighborsClassifier
#importing accuracy_score class from sklearn.metrics
from sklearn.metrics import accuracy_score

#declare the value of k
k = 3

#instantiating an object of KNeighborsClassifier with above k value
knn_modal = KNeighborsClassifier(n_neighbors=k)

#training the model using the training data and training labels
knn_modal.fit(X_train, y_train)

#predecting the testing data
y_predict = knn_modal.predict(X_test)

#calculating the accuracy of the modal
accuracy = accuracy_score(y_test, y_predict)

print(accuracy)

0.6973684210526315


In [22]:
#Logistic Regression

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=4)

my_logreg = LogisticRegression()
# Training ONLY on the training set:
my_logreg.fit(X_train, y_train)
y_predict_lr = my_logreg.predict(X_test)
score_lr = accuracy_score(y_test, y_predict_lr)
print(score_lr)

0.7236842105263158


C:\Users\patel\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


In [23]:
#Decision Tree (with random_state=5)

my_decisiontree = DecisionTreeClassifier(random_state=5)
# Training ONLY on the training set:
my_decisiontree.fit(X_train, y_train)
y_predict_dt = my_decisiontree.predict(X_test)
score_dt = accuracy_score(y_test, y_predict_dt)
print(score_dt)

0.6052631578947368


so that we can see that new accuracy are as below

Accuracy of the modal
0.6973684210526315

#Logistic Regression
0.7236842105263158

#Decision Tree (with random_state=5)
0.6052631578947368

### h- 
Now, repeat part (e) with the new dataset that you built in part (f), but this time using
Cross-Validation. Thus, rather than splitting the dataset into testing and training, use
10-fold Cross-Validation (as we learned in Lab4) to evaluate the classification methods
and report the final prediction accuracy. 

In [24]:
# importing the method:
from sklearn.model_selection import cross_val_score


In [25]:
# Applying 10-fold cross validation with "logistic regression" classifier:

# function cross_val_score performs Cross Validation:
accuracy_list = cross_val_score(my_logreg, X, y, cv=10, scoring='accuracy')

print(accuracy_list)

[0.70967742 0.70967742 0.64516129 0.83333333 0.83333333 0.53333333
 0.73333333 0.73333333 0.65517241 0.72413793]


C:\Users\patel\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\patel\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\patel\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\patel\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\patel\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence t

In [26]:
# use average of accuracy values as final result
accuracy_cv = accuracy_list.mean()

print(accuracy_cv)

0.711049314052651
